catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,445 @@
+  }
 
-    runner.shutdown();
+  /**
+   * Sets the {@link FederatedSource}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
+   *
+   * @throws IllegalArgumentException if {@code federatedSource} is {@code null}
+   */
+  public void setFederatedSources(final List<FederatedSource> federatedSources) {
+    notNull(federatedSources);
+
+    LOGGER.trace(
+        "Setting federated sources: old size={}, new size={}",
+        this.federatedSources.size(),
+        federatedSources.size());
+    this.federatedSources = federatedSources;
+  }
 
-    handle.cancel(true);
+  /**
+   * Sets the {@link CatalogProvider}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
+   *
+   * @throws IllegalArgumentException if {@code catalogProvider} is {@code null}
+   */
+  public void setCatalogProviders(final List<CatalogProvider> catalogProviders) {
+    notNull(catalogProviders);
 
-    scheduler.shutdownNow();
+    LOGGER.trace(
+        "Setting catalog providers: old size={}, new size={}",
+        this.catalogProviders.size(),
+        catalogProviders.size());
+    this.catalogProviders = catalogProviders;
   }
 
   /**
-   * Start method for this poller, invoked by the CatalogFramework's blueprint when the framework is
-   * registered/installed. No logic is executed except for logging the framework name.
+   * Sets the {@link CatalogStore}s of which to check the availability periodically according to
+   * {@link #setPollIntervalMinutes(long)}
    *
-   * @param framework the catalog framework being started
-   * @param properties unused, but required by blueprint
+   * @throws IllegalArgumentException if {@code catalogStore} is {@code null}
    */
-  public void start(CatalogFramework framework, Map properties) {
-    LOGGER.debug("Framework started for [{}]", framework);
+  public void setCatalogStores(final List<CatalogStore> catalogStores) {
+    notNull(catalogStores);
+
+    LOGGER.trace(
+        "Setting catalog stores: old size={}, new size={}",
+        this.catalogStores.size(),
+        catalogStores.size());
+    this.catalogStores = catalogStores;
   }
 
-  public int getInterval() {
-    return interval;
+  /**
+   * This method is used to keep the cache up to date.
+   *
+   * <p>The current {@link Source}s configured by {@link #setConnectedSources(List)}, {@link
+   * #setFederatedSources(List)}, {@link #setCatalogProviders(List)}, and {@link
+   * #setCatalogStores(List)} are discovered.
+   *
+   * <p>A process to check (or recheck) the availability of each {@link Source} is started unless
+   * the {@link Source} was created or modified within the last poll interval and the availability
+   * check triggered by that action (see {@link #event(ServiceEvent, Map)} has not yet completed.
+   *
+   * <p>TODO Should this method be synchronized?
+   */
+  private void handleAllSources() {
```

Comment:
❗️ If there are any unhandled exceptions in this method, all future calls will be suppressed.

>If any execution of the task encounters an exception, subsequent executions are suppressed.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit)

Here is an example of how we can protect against this.
https://github.com/codice/ddf/blob/master/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L88-L92

Reply:
Not sure if I added this correctly
```java
scheduledExecutorService.scheduleAtFixedRate(
    () -> {
      try {
        sourcePoller.pollAllSources(pollInterval, pollIntervalTimeUnit);
        // If there are any unhandled exceptions in this method, all future calls will be
        // suppressed.
      } catch (VirtualMachineError vme) {
        throw vme;
      } catch (Throwable throwable) {
        LOGGER.warn(
            "Scheduled SourcePoller poll failed unexpectedly. Try restarting", throwable);
      }
    },
    0,
    pollInterval,
    pollIntervalTimeUnit);
```